doc/config-generic-update-rollback/Json_Patch_Ordering_using_YANG_Models_Design.md

```diff
@@ -0,0 +1,698 @@
+  
+# 3 Design
+
+## 3.1 Overview
+A small modification to the [DFS algorithm](##2.2-Functional-Description) listed before is introducing memoization, where the function is optimizing for the shortest number of moves.
+```python
+def order-patch(patch):
+  current_config = get_current_config()
+  target_config = simulate_patch(patch, current_config)
+  return rec(current_config, target_config)
+
+visited = {}
+mem = {}
+def rec(current_config, target_config):
+  if current_config == target_config:
+      return []
+  if current_config in mem: # If already calculated, just return the result form memory
+      return mem[current_config]
+  if current_config in visited:
+      return None
+  visited[curr_config] = True
+
+  moves = get_all_moves(current_config)
+  best_moves = None
+  for move in moves:
+      if is_valid(move, current_config):
+          new_config = apply_move(move, current_config)
+          new_moves = rec(new_config, target_config)
+          if new_moves != None:
+              # Minimize on the number of moves
+              if best_moves == None or len(best_moves) > 1 + len(new_moves):
+                  best_moves = [move] + [new_moves]
+  
+  mem[current_config] = best_moves
+  return best_moves
+```
+
+This optimization can help reduce the number of result JsonChanges, and ultimately reduce the time needed to completely apply all the JsonChanges.
+
+SONiC YANG models are going to be used for the following:
+- Validating config state after applying a move
+- Generating other moves
+- Host flags/configs for move validation
+
+### 3.1.1 Using YANG Models to Validate Config State
+This is a straightforward task and can be done using [SONiC Libynag wrapper](https://github.com/Azure/sonic-buildimage/blob/master/src/sonic-yang-mgmt/sonic_yang.py), specifically `validate_data_tree`.
+
+### 3.1.2 Using YANG models to Generate Other Moves
+Check [2.2.3 Generating all possible moves](#223-generating-all-possible-moves) to learn more about what are the possible moves.
+
+YANG models can help with:
+- Lines getting removed which have references, can be substituted by deleting their references first as it is required anyway (check [2.2.3.3 Generating Other Moves](#2233-generating-other-moves) for details). Getting the dependency of some data can be done using [SONiC Libynag wrapper](https://github.com/Azure/sonic-buildimage/blob/master/src/sonic-yang-mgmt/sonic_yang.py), specifically `find_data_dependencies`.
+- Any other moves should be generated using YANG models, if YANG models data are not enough an extension to YANG models can be added to help with generating new moves.
+
+### 3.1.3 Using YANG models to Host Flags for Move Validation
+#### 3.1.3.1 Extension: create-only
+Some fields are not allowed to be modified and can only be created, such as `/PORT/Ethernet0/lanes` which is marked as `CREATE_ONLY` in [opencomputeproject/SAI/blob/master/inc/saiport.h](https://github.com/opencomputeproject/SAI/blob/master/inc/saiport.h#L624):
```

Comment:
Any idea if we can mark in Yang model that a particular object is create-only,  I noticed the extension down later

Reply:
Yes, will use an extention.